apis/v1alpha2/grpcroute_types.go

```diff
@@ -0,0 +1,552 @@
+// evaluate to true only if all conditions are satisfied.
+//
+// For example, the match below will match a gRPC request only if its service
+// is `foo` AND it contains the `version: v1` header:
+//
+// ```
+// matches:
+// - method:
+//    type: Exact
+//    service: "foo"
+//   headers:
+//   - name: "version"
+//     value "v1"
+// ```
+type GRPCRouteMatch struct {
+	// Path specifies a gRPC request service/method matcher. If this field is not
+	// specified, all services and methods will match.
+	//
+	// +optional
+	// +kubebuilder:default={type: "Exact"}
+	Method *GRPCMethodMatch `json:"method,omitempty"`
+
+	// Headers specifies gRPC request header matchers. Multiple match values are
+	// ANDed together, meaning, a request MUST match all the specified headers
+	// to select the route.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []GRPCHeaderMatch `json:"headers,omitempty"`
+}
+
+// GRPCMethodMatch describes how to select a gRPC route by matching the gRPC
+// request service and/or method.
+//
+// At least one of Service and Method MUST be a non-empty string.
+type GRPCMethodMatch struct {
+	// Type specifies how to match against the service and/or method.
+	// Support: Core (Exact with service and method specified)
+	//
+	// Support: Implementation-specific (Exact with method specified but no service specified)
+	//
+	// Support: Implementation-specific (RegularExpression)
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *GRPCMethodMatchType `json:"type,omitempty"`
+
+	// Value of the service to match against. If left empty or omitted, will
+	// match all services.
+	//
+	// At least one of Service and Method MUST be a non-empty string.
+	// +optional
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:Pattern=`^[^\/]*$`
+	Service *string `json:"service,omitempty"`
+
+	// Value of the method to match against. If left empty or omitted, will
+	// match all services.
+	//
+	// At least one of Service and Method MUST be a non-empty string.
+	// +optional
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:Pattern=`^[^\/]*$`
+	Method *string `json:"method,omitempty"`
+}
+
+// MethodMatchType specifies the semantics of how gRPC methods and services are compared.
+// Valid MethodMatchType values are:
+//
+// * "Exact"
+// * "RegularExpression"
+//
+// Exact paths MUST be syntactically valid:
+//
+// - Must not contain `/` character
+//
+// +kubebuilder:validation:Enum=Exact;RegularExpression
+type GRPCMethodMatchType string
+
+const (
+	// Matches the method or service exactly and with case sensitivity.
+	GRPCMethodMatchExact GRPCMethodMatchType = "Exact"
+
+	// Matches if the method or service matches the given regular expression with
+	// case sensitivity.
+	//
+	// Since `"RegularExpression"` has implementation-specific conformance,
+	// implementations can support POSIX, PCRE, RE2 or any other regular expression
```

Comment:
I think we just need to set expectation and make it easier for the user and the implementer. BRE afiak is what is supported across all implementation and covers all the use cases the extended stuff are just basic implementation done differently (as in there is no additional feature or to be more precise it does the same things using smaller number of characters).